specify/specify7
# Locality and Geocoorddetail batch import tool

Fixes #4319

The Locality and GeoCoordDetail batch import tool allows users to match and update the Latitude1, Longitidude1, and Datum fields for Locality records and associate new GeoCoordDetails for the matched Locality records. 

The primary motivation for the tool is such that the georefrencing for existing records in the database can take place outside of Specify (such as with [CoGe](https://coge.geo-locate.org/About.htm)), and there is no easy way to currently repatriate the data back into Specify and update those records. 

This process is facilitated by requiring the guids of the Locality records in the CSV. Specify will use these guids to match/find the corresponding Locality record(s). 
With the Locality record, the tool does the following: 
- If any `latitude1`, `longitude1`, and/or `datum` fields have values for the Locality record in the CSV, Specify will overwrite the current values in the database with those specified in the CSV
- If any other column in the CSV matches a non-relationship field name in the GeoCoordDetail table, Specify will delete the previous GeoCoordDetail associated with the Locality record and create a new one populated with field values from the CSV
- If there are any other columns in the CSV, Specify will display a warning to the user indicating there are unknown columns, but allow the user to proceed; the unknown columns will be ignored if the user proceeds. 

---
### Demo
Currently, access to the tool can be found in the User Tools menu: 

<img width="1436" alt="locality_import_button" src="https://github.com/specify/specify7/assets/64045831/20c8556e-35c3-4b71-a6e7-2f82ba4b5de2">


The primary interface is very similar to that of importing a file to be used as a DataSet in the WorkBench. Specify supports different encodings and delimiters, and displays the first 100 rows of the file:

<img width="1427" alt="coge_preview" src="https://github.com/specify/specify7/assets/64045831/a3181a88-26be-4a1c-bf07-33461e7b9607">

The `Import File` button can be pressed to initiate the parsing/upload. 

If there are any unknown columns in the dataset, a warning will first be displayed to the user so they can decide how to proceed: 

<img width="1427" alt="coge_column_warning" src="https://github.com/specify/specify7/assets/64045831/6d6e4b1a-9b5e-43d0-b5f8-f6e910cfc54a">

If there are any bad values discovered when parsing the results, Specify will display a downloadable message stating which line the error was found on along with a message with more information about how to resolve the parsing error:

<img width="1436" alt="parse_error" src="https://github.com/specify/specify7/assets/64045831/93576706-98ce-4b26-ac2c-31d4511fa0a0">


If the parsing and upload is successful, Specify will display a brief results page stating how many records of each table were affected, along with the option to create a RecordSet of modified Locality Records:

<img width="1436" alt="upload_results" src="https://github.com/specify/specify7/assets/64045831/3e813a55-28fb-44d2-ae53-ce40ed310ef3">

--- 

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions

For testing purposes, here is an example set of data for use on the `KUFish` database: 
[kufish_coge_sample.csv](https://github.com/specify/specify7/files/15257843/kufish_coge_sample.csv)

Other datasets have been made for `FITZ_NHMD`, `fwri`, `KUBirds`, and `lsumzmammals` and are available in the following folder in the google drive: 
https://drive.google.com/drive/folders/1B0hKQMBaX82nBUOB95uEa3H0MIrGrBv7?usp=drive_link

A foundation for other Datasets for use in other databases can be made by creating a Locality -> GeoCoordDetail query and exporting the results. For an example, here is the Query used to populate most of `kufish_coge_sample.csv`: 
https://kufish51623-edge.test.specifysystems.org/specify/query/191/

There should be no errors in the `kufish_coge_sample` dataset presently, so modifications can be made to accomplish the desired errors and edge cases. 

Generally, the cases which would need to be tested are: 
- [ ] Ensure an error message is displayed to the user when there is no Locality which has a guid which exists in the dataset
- [ ]  Data in unknown columns is properly ignored and not accounted for when uploading
- [ ] If a Locality contains existing data for `GeoCoordDetails`, it should be deleted if there is at least one `GeoCoordDetail` field in the dataset
- [ ] If there are no values in any GeoCoordDetail field for a Locality in the dataset and the Locality contains a GeoCoordDetail in Specify, the GeoCoordDetail should not be overwritten
- [ ] Ensure that only the Locality fields which contain data in the dataset get updated for a Locality in the dataset (i.e., if `longitude1` and `datum` have values in the dataset but `latitude1` is either not a column or empty, ensure that `latitude1` is never overwritten) 
- [ ] Ensure parsing error messages are intuitive enough to diagnose and resolve the problem with the dataset
  - Potential ways to introduce invalid values include: 
    - Entering a value in a field which exceeds the maximum length for the field (the maximum length for a field can be found in the Schema Configuration) 
    - Entering an incorrect type of value for the field (such as entering a letter into a number/integer field)
    - Not following a UIFormatter for a field
